package world_utils.msgs;

import "pose.proto";
import "vector3d.proto";

message WorldUtilsRequest
{
    /* @brief Type of request */
    enum Type{
        /* @brief Spawn entity in simulation */
        SPAWN   = 1;
        /* @brief Move entity in simulation */
        MOVE    = 2;
        /* @brief Remove entity from simulation  */
        REMOVE  = 3;
        /* @brief Start or stop physics simulation */
        PHYSICS = 4;
        /* @brief Pause or resume world */
        PAUSE   = 5;
        /* @brief Request entity or world information */
        STATUS  = 6;
    }

    /* @brief Type of model to be spawned */
    enum ModelType{
        /* @brief Spawn sphere in simulation */
        SPHERE          = 1;
        /* @brief Spawn cylinder in simulation */
        CYLINDER        = 2;
        /* @brief Spawn cube in simulation */
        BOX             = 3;
        /* @brief Spawn custom object from sdf string */
        CUSTOM          = 4;
        /* @brief Spawn custom light from sdf string */
        CUSTOM_LIGHT    = 5;
        /* @brief Spawn a model included in gazebo model path */
        MODEL           = 6;
    }

    /* @brief Type of request */
    required Type                   type            = 1;
    /* @brief Model type for spawn */
    optional ModelType              model_type      = 2;
    /* @brief Model name */
    optional string                 name            = 3;
    /* @brief Model pose */
    optional gazebo.msgs.Pose       pose            = 4;
    /* @brief Model mass */
    optional double                 mass            = 5;
    /* @brief Material script uri */
    optional string                 texture_uri     = 6;
    /* @brief Material script name */
    optional string                 texture_name    = 7;
    /* @brief radius of a cylinder or a sphere */
    optional double                 radius          = 8;
    /* @brief Length of a cylinder */
    optional double                 length          = 9;
    /* @brief Dimensions of a box */
    optional gazebo.msgs.Vector3d   box_size        = 10;
    /* @brief Custom object sdf string */
    optional string                 sdf             = 11;
    /* @brief World physics state */
    optional bool                   state           = 12;
}